### ISSUE TITLE
Community feedback needed for project's development guidelines.

### ISSUE THREAD

**Issue by fmv1992** [python-mode/python-mode#805]
Dear python-mode community,

I've been participating on this project in the background whilst developing my
own views of pymode at my own fork.

I've seen some good issues and points being brought up occasionally by other
users and I think this is a good oportunity to settle on some 'rules' and
practices for this project. The idea is to provide clearer guidelines on
contributing, future directions, etc.

So now its the time to speak up.

After our conclusions on this topic I'll commit myself to be more active and
assertive in closing/accepting pull requests.

So from my part I say we must have some sort of testing before accepting PRs.
I will put a draft on this topic later on. You can check it out here:

* https://github.com/python-mode/python-mode/blob/develop/tests/test.sh

I'm marking people with recent activity below:

* @AdibMurshed
* @ConradP
* @MadWombat
* @alphaCTzo7G
* @bbbco
* @brycepg
* @diraol
* @duytrung
* @fdev31
* @hawkeyej
* @icemanx7
* @kerwinxu
* @klen
* @moorepants
* @olalonde
* @saltycraig
* @smheidrich
* @steven-walton
* @wklm

# Conclusions

## PRs

* Should be tested before accepted (open to discussion on methodology).

## Python code

* Should conform to PEP8 and PEP257 (open to discussion).

## Important feature requests

* Priority:
    * Improve python-mode's communication (including our one-liner slogan and our
    `readme.rst`).
        * Redesign our slogan (make it more IDE-like).
        * Redesign our `readme.rst`.
    * Move third party libraries to submodules.
        * Advantages:
            * Provide modularity for our code base.
            * Moves and unnecessary burden of keeping the code base of different
            projects up to date.
            * Allow users to handpick their preferred versions of each library.
        * Disadvantages:
            * Forces users to issue and additional git command.
    * Cleanup libraries and huge codebase on `pymode/libs`.
    * Create a roadmap of python libraries that we want to include.
        * We should have a clear line on what to include and not to include,
        specially because adding up libraries probably would give diminishing
        returns (there is some overlap between them).
* Non priority:
    * Find a suitable library for completion (probably jedi).
    * Add support for https://github.com/google/yapf.

## Contributors

* Permanent contributors should be added after some time of steady
  participation on this project (clarification needed).

**Comment by fmv1992** [comment]
Dear community,

I'm pleased to announce commit bf018e262a0601bf5c9290e02d038f4694ff03cc.

The most important changes are that libraries found in `pymode/libs` and other
files were moved to submodules.  
Other minor improvements were also executed.

This greatly improves the maintenance in our project as thousand-lines long
commits are not necessary. To update all submodules to their latest version a
single command must be issued:

`git submodule update --recursive --remote`

I strongly the community to check out the new `dev_submodules` branch and test
it. If you do so do not forget to execute the following command:

**`git submodule update --recursive --remote`**

I hope to hear feedback from you soon!

Best,

Felipe.

PS: Just for the record, the 'legacy' commit is
b848f6294ca5858fd17eaf14783fa2042e8142b7.

**Comment by MadWombat** [comment]
The only problem I see with using submodules is with Vim package managers like Vundle. Right now all that is required to make python-mode work is to add a Plugin line in vimrc and run PluginInstall. With submodules, one would have to go into the cloned repo and run submodule init and update commands by hand.

**Comment by alphaCTzo7G** [comment]
I have been using python-mode, with the default configuration for a couple of weeks. Thanks for creating this.

Couple of suggestions which may make it more popular and lead to continued development and improvements:

1. I agree, that we should move towards simplicity, rather than ask the new user to learn more things before being able to use pymode. There are some arguments against making VIM into an IDE, but pymode, is kindof a python IDE pluggable into VIM, and thats why I find it amazing. I am guessing a lot of people, start using pymode because they miss the "integrated" development that an IDE provides, but also want to use VIM (and not the pluging for VIM provided by the IDE). Good software provides an abstraction for the underlying complexity. Further, the tagline reads "Vim python-mode. PyLint, Rope, Pydoc, breakpoints from box." Unless you have been using Python for a while, people don't usually know what these terms mean. Even though Pycharm probably uses some of these libraries, or something similar in the background, they dont say anything about these.

Why don't we advertise it as such, such as "Python IDE for VIM"?

2. The number of contributors to a project like this will be proportional to the number of users. With more user-friendly UIs like Pycharm, most new Python users are going to Pycharm. I wonder why we have such a lengthy first page. This would lead to the "paradox of choice" https://en.wikipedia.org/wiki/The_Paradox_of_Choice. Here's the launch page for Tensorflow https://github.com/tensorflow/tensorflow. I am pretty sure Tensorflow is way more complicated than python-mode. Maybe we can list the high-level functionalities provided by Pymode with some videos/demonstrations, and leave the rest to the wiki?

3. Neovim has moved to Apache 2.0 license. I have heard that a lot of companies prevent/discourage use of GPL licensed products, because of IP risk. http://corporate.findlaw.com/business-operations/the-risks-of-open-source-software.html.  This is probably one of the reasons Neovim is changing their license. I would guess most programmers work in big corporations, and being GPL licensed, we automatically lose a big chunk of users. https://github.com/neovim/neovim/blob/master/README.md#license. Can we move to Apache 2.0/MIT license? 

4. I notice that there are multiple libraries being used, some of which maybe redundant, not developed anymore or maybe there are better options available since 2015. http://liuchengxu.org/posts/use-vim-as-a-python-ide/. Can we use some of the more mordern ones?



In general, I have seen that VIM has tonnes of plugins for the same function, and it takes a while to figure out which one maybe best suited. Where python-mode could really add value for Python coders in VIM, is to screen all these plugins and provide, the best ones, great defaults, such that the user can get a IDE like environment quickly and get back to coding.

Thanks a lot for creating pymode. I can help with some of these.. or other ideas from the community over time.

**Comment by fmv1992** [comment]
@madwombat,

**tl;dr: submodules structure benefits outweighs its costs.**

Those shortcomings are known but I think the benefits outweighs the costs in
this case. To update a single library usually means removing one folder and
replacing with another one full of changes that are tracked (and shouldn't be).
And this a mice and rat game because we will keep having to "paste" libraries
into our folder.

Another point is that in the makefile we are installing those libraries via pip
which I also think is not the best approach. Vim should not interfere with the
user's libraries.

And to remedy this issue one command is enough:
`git submodule update --init --recursive && git submodule update --recursive
--remote`

My idea is to move the tip of this branch to `master` (thus keeping the current
layout) and change `develop` to this new structure and keep it this way for 1-3
months to improve user testing and collect feedback. Also the libs folder is
overcomplicated now and I suspect we don't use a good part of the code that is
there.

**Comment by fmv1992** [comment]
@alphaCTzo7G,

Those are very good points!

Brief comment:

1. Totally agree on this improvement of the plugin. Good point.

1. (Same argument as 1.) Redesign of the `readme.rst` is needed.

1. This point is hairy I think for two reasons.  
   First I don't think we can change the license without the author's consent
   (I'm not the author).  
   Second I don't think this software clashes with corporate interests/IP
   because you are only required to release source code if you distribute this
   software as part of your product. In other words, corporations can use this
   software to develop proprietary python applications and distribute them with
   whatever licenses. If they distribute their product with embedded
   python-mode in it then they have to distribute the source code too (which
   I hardly think is the case).

1. I will put this in our roadmap. The libs folder is still obscure to me (in
   terms of what is actually used and what is not) and mapping them is
   a priority.

Great feedback! And I also thank those who created python mode :)

**Comment by songww** [comment]
python-mode is a great. Thanks for your work.
There are many plugins for flake8, but submodule or sub-folder is too hard to use that.
When I register flake8 to git hooks, it will has two copies(maybe version also is different).
Use the user's library as an option？

**Comment by fmv1992** [comment]
@songww,

If I understood correctly you also oppose the use of submodules.

Again I think that the benefits outweigh the costs as a single command has to
be issued and the user can either decide which version of each library he wants
to use.

Also I don't want vim to interfere with a users choice of python packages (pip
packages).

Flake8 sounds like a good idea. But before we start adding more
packages/complexity to our mix I think we should focus on what we have right
now.

Thus I suggest that we create a list of linters/tools and discuss on those that
should be implemented. But again our starting point should be what we have
right now and have them operating correctly.

Good feedback!

**Comment by fmv1992** [comment]
The most pressing matter is rewriting the `readme` file I guess.

I'll give it a try on 2017/11/15 ; by no means wait for me to improve it!

**Comment by alphaCTzo7G** [comment]
I will work on it a bit over the next few days.. I have been going over the options over the weekend quite a bit: 

1. There have been a lot of interesting developments going on in the development of "Language server protocols", which might unite the need for separate libraries for separate editors and languages.. but I am not sure how far out that is: https://github.com/prabirshrestha/vim-lsp

2. It seems that almost universally people are using jedi instead of rope for autocompletion

3. In terms of linting, pep257 has be renamed/upgraded to pydocstyle. 

4. There's ide's based on Neovim;https://github.com/onivim/oni

5. for formatting, Google/yapf seems to be more popular than autopep8.

However, none of these are well integrated like python-mode. You have to do all of it on a one by one basis..

**Comment by fmv1992** [comment]
@alphaCTzo7G,

I'll go through each point later but something occurred to me.

Assume the following are true:

1. Most users of vim use it to code in more than one programming language.
1. Advanced users of vim are aware of its code completion features and plugins.

Then I think that the following is a conclusion: each user will setup its own
code completion tool.

Thus I think that we can in the future add a robust completion feature that
should be off by default (let's go with jedi if what you said is true) but for
now assume that the users will have their own completions set up.

For instance I have tried both NeoComplete
<https://github.com/Shougo/neocomplete.vim> and YouCompleteMe
<https://valloric.github.io/YouCompleteMe/> and approved both.

The idea behind 'outsourcing' this is to let us focus on the most pressing
needs that we have right now (the `readme.md`, the roadmap for needed
maintenances and then adding new features IMO; adding support for jedi being
part of the third group).

See a related issue here:

https://github.com/python-mode/python-mode/issues/806

**Comment by alphaCTzo7G** [comment]
I agree we should add tests. 

We can start with something simple like you had. Havent use these myself, but there are also some vim testing frameworks as well:

1. https://stackoverflow.com/questions/3029882/tools-for-testing-vim-plugins
2. https://github.com/janko-m/vim-test
3. http://vim.wikia.com/wiki/Integration_with_PyUnit_testing_framework
4. http://kana.github.io/config/vim/vspec.html
5. https://github.com/dhruvasagar/vim-testify
6. https://devhub.io/zh/repos/junegunn-vader.vim

**Comment by alphaCTzo7G** [comment]
In terms of modularity, we should probably try to delegate as much of the work as possible to third party plugins. That way we can reduce our work, and as and when those plugins improve, we can just pull the improvements into this one.

Before we decide for how to modify this library, so that its more useful for everybody and so that we can sustain it long term, should we come up with a list of features/specs that are necessary to be able to handle large projects?

Perhaps, we can chose a reasonably large python project OS project/projects (sklearn, tensorflow are pretty large and could be good test vehicles), and then try to figure out what plugins are necessary to be able to handle this project and meet the specs.

**Comment by fmv1992** [comment]
I think so as well. I'm not sure the best way to track those needs though.

**Comment by fmv1992** [comment]
I have reworked the mechanism and output of troubleshooting (updated the readme.md and the documentation as well).

This will possibly make it even more helpful to diagnose problems.

**Comment by fmv1992** [comment]
1. There have been a lot of interesting developments going on in the development of "Language server protocols", which might unite the need for separate libraries for separate editors and languages.. but I am not sure how far out that is: https://github.com/prabirshrestha/vim-lsp
    * Sounds good but I don't quite grasp the tangible aspect of it.

2. It seems that almost universally people are using jedi instead of rope for autocompletion
    * Accepted; already on top post.

3. In terms of linting docstrings, pep257 has be renamed/upgraded to pydocstyle. 
    * This goes away with `dev_submodules` merging.

4. There's ide's based on Neovim;https://github.com/onivim/oni
    * I think that the idea in the long run is to have this IDE-like
      experience. However I'm not sure what the trade-offs are and I have not
      tried neovim (nor oni).  
      And well, this is Vim, so when faced with any trade-offs between 'bells
      and whistles' and vim's 'culture' so to speak we should go the vim-way.

5. for formatting, Google/yapf seems to be more popular than autopep8.
    * I have not heard about it so far but it looks just great. Added to top
      post!

6. a lot of the complaints regarding using pymode vs IDEs is that the lack of
   async in VIM 7.4 makes it slow to manage large libraries. So people switch
   to Pycharm.
    * Vim is not for everyone. And I don't think that this is a problem, nor
      a matter of 'better vs worse'. It is just a matter of style. It sure does
      not mean that we rejoice in having things barebones-vim-like. We should
      go after improvements and functionality but we will never be a GUI-IDE.
      Let's keep this in mind.

7. For refactoring, rope is pretty much all there is for VIM.
    * Then we go with Rope ;)

* * *

Two weeks have passed since the opening of this post. I think we are able to
go with submodules, freeze this discussion for a while and work on this small
roadmap that we have agreed upon.

What do you think?

**Comment by fmv1992** [comment]
Dear python-mode community,

I'm super glad to announce that we now have a test suite up and running!

More important than having the Travis CI badge:

![What a nice badge huh?!](https://travis-ci.org/python-mode/python-mode.svg?branch=develop)

Is to impart reliability to our software by creating a **meaningful** test
suite.

We now have tests that despite its complicatedness test python-mode's most
important functions.

The documentation also got a new section on Development.

None of those additions are set in stone but they represent a milestone in our
projects governance.

Best,

**Comment by diraol** [comment]
Hi folks,

A little bit late on the discussion (as I've told @fmv1992, I was traveling on the last weeks)...

What about using git subtree instead of git submodules?
http://endot.org/2011/05/18/git-submodules-vs-subtrees-for-vim-plugins/

I haven't used it my self, but it looks like a nice way to manage 'external dependencies'.

This would add some extra effort to the development community (but not the craziness of copying files around), make the git history a little bit dirtier, but would ease the life of our users, avoiding the need of using extra git commands to have the plugin working.

Does it make any sense?

**Comment by fmv1992** [comment]
@diraol,

I cannot access this website right not but from I read on:

1. https://stackoverflow.com/questions/31769820/differences-between-git-submodule-and-subtree

I think submodules is what we are looking for. We already have 6 or 7 plugins
and this number is likely to increase in the future. As we have seen those
changes from one version to another can be several hundreds of lines long.

The conclusion that I have is that if I want to track pymode's file I'll have
a lot of difficulty.

One command can sometimes be a hindrance but I think it still wins over
subtrees.

We could try that on a different branch in the future and see how it goes :)
